Change maxuploadtarget recommended minimum calculation

jonasschnelli opened the pull request:
Addresses #8555.

Changes the recommended `-maxuploadtarget` minimum from 144 blocks per day to 72 blocks per day. With the SWs `MAX_BLOCK_SERIALIZED_SIZE` of 4MiB, 144 blocks per day would result in a recommended minimum of 576MiB per day = ~17GB per Month in uploading blocks. IMO this is too hight for a recommended _minimum_.

This PR changes the minimum calculation from 144 to 72 blocks per day ("half day of full blocks").
Anyway, the calculation is just a form of an vague estimation of how many MiB in historical blocks we should recommend to broadcast upstream when `-maxuploadtarget` is set. It does not take -maxconnection etc. into the calculation.

~~The current maxuploadtarget recommendation is calculated using `MAX_BLOCK_SERIALIZED_SIZE` (4MB) instead of `MAX_BLOCK_BASE_SIZE` (1MB). Given that the function is there to reduce uploading historical blocks, using `MAX_BLOCK_BASE_SIZE` makes more sense.~~

jonasschnelli committed:
Fix maxuploadtarget recommended minimum calculation

MarcoFalke commented on src/net.cpp:
```diff
@@ -2193,7 +2193,7 @@ void CNode::RecordBytesSent(uint64_t bytes)
+    uint64_t recommendedMinimum = (nMaxOutboundTimeframe / 600) * MAX_BLOCK_BASE_SIZE;
     nMaxOutboundLimit = limit;
 
     if (limit > 0 && limit < recommendedMinimum)
```
Nit: Can you also adjust the LogPrintf below to display the size in whatever unit `-maxuploadtarget` is.

jonasschnelli replied:
@MarcoFalke: I can't parse your sentence. :)
Right now we have: `LogPrintf("Max outbound target is very small (%s bytes) and will be overshot. Recommended minimum is %s bytes.\n", nMaxOutboundLimit, recommendedMinimum);` which prints the desired output limit (1:1 from `-maxuploadtraget` and the recommended minimum, both are stated to be in bytes.

MarcoFalke replied:
Huh, I thought -maxuploadtarget was in MB or MiB?

jonasschnelli replied:
Yes. It is in MB. But the LogPrintf does print out the passed in MB values in bytes (`nMaxOutboundLimit` is `-maxuploadtarget *1024 *1024`)

MarcoFalke replied:
I think it would be more user friendly to print this out in MiB, so the user can see what value for `-maxuploadtarget` should be used instead.

MarcoFalke commented on src/net.cpp:
```diff
@@ -2193,7 +2193,7 @@ void CNode::RecordBytesSent(uint64_t bytes)
 {
     LOCK(cs_totalBytesSent);
-    uint64_t recommendedMinimum = (nMaxOutboundTimeframe / 600) * MAX_BLOCK_SERIALIZED_SIZE;
+    uint64_t recommendedMinimum = (nMaxOutboundTimeframe / 600) * MAX_BLOCK_BASE_SIZE;
```
Though, `MAX_BLOCK_SERIALIZED_SIZE` is your cost "on the wire". This will be true after segwit is active.

I'd rather keep this line as is and adjust the documentation.

jonasschnelli replied:
Right.. the recommended minimum must not reflect the historical blocks sizes, it should reflect the current blocks. Should't this be switched from `MAX_BLOCK_BASE_SIZE` to  `MAX_BLOCK_SERIALIZED_SIZE ` when SW activates?

MarcoFalke replied:
But keep in mind that updating the doc is required in either case: If you keep `MAX_BLOCK_SERIALIZED_SIZE` or if you add the condition.

jonasschnelli commented:
Changed the recommended minimum "calculation" (it's not really a calculation) from 144 blocks to 72 full blocks (on the wire size) per day.
Not doing this, would result in a recommended upload minimum (after SW has been deployed) of ~17GiB per month (which is too high IMO for a *minimum*).

Updated the docs, removed the MiB values.

MarcoFalke commented:
Thanks, looks better! utACK 5e872003e7243c4e02eec4f57f3a001dd20d925c

wodry commented:
Feedback from a rather "normal user":

In reduce-traffic.md would now be written:
_"1. Use -maxuploadtarget=<MiB per day> [...] The current recommended minimum is 72 full blocks (by size "on the wire" per day)""_

In my opinion, this is not understandable for a normal user- what value should that be in MiB ??

I think it would be better to recommend a concrete value in the unit of the parameter (MiB), and adapt that recommendation with increasing population of Segwit (and therefore the increasing ammount of bytes stored in one block, as I understand?) accross the upcoming (minor) releases.

At least, if the current text should remain, there should be more explanation of how could one calculate _"72 full blocks (by size "on the wire" per day)""_ to MiB.

MarcoFalke commented on doc/reduce-traffic.md:
```diff
@@ -19,8 +19,8 @@ This is *not* a hard limit; only a threshold to minimize the outbound
-historic blocks. **The recommended minimum is 144 blocks per day (max. 144MB
-per day)**
+historic blocks.
+**The current recommended minimum is 72 full blocks (by size "on the wire" per day)**
```
Nit: (to address the feedback by @wodry )

**The current recommended minimum is `-maxuploadtarget=$recommendedMinimum`, assuming an average serialized block size of 2MB.**

MarcoFalke commented on src/net.cpp:
```diff
@@ -2193,11 +2193,12 @@ void CNode::RecordBytesSent(uint64_t bytes)
 {
     LOCK(cs_totalBytesSent);
-    uint64_t recommendedMinimum = (nMaxOutboundTimeframe / 600) * MAX_BLOCK_SERIALIZED_SIZE;
+    // set the recommended minimum to ~72 full blocks per day (maximal upstream full blocks worth of a ~1/2 day)
```
Nit: (as above)

```cpp
// assume the average size of serialized blocks is half the maximum

jonasschnelli commented:
Nits addressed.

sipa commented:
utACK

MarcoFalke commented:
utACK d2ce6d3

wodry commented:
Instead of
_"The current recommended minimum is 288MiB, assuming an average serialized block size of 2MB."_
I would suggest
_"The current recommended minimum is 288 MiB, assuming an average serialized block size of 2 MiB."_
, to have a consistent unit of MiB and correct spelling.

MarcoFalke commented:
@wodry The block size limit is denoted in the "decimal" MB (Like most other sizes in Bitcoin Core). Unfortunately some palaces use MiB, but there is nothing you can do about it.

MarcoFalke commented on doc/reduce-traffic.md:
```diff
@@ -19,8 +19,8 @@ This is *not* a hard limit; only a threshold to minimize the outbound
-historic blocks. **The recommended minimum is 144 blocks per day (max. 144MB
-per day)**
+historic blocks.
+**The current recommended minimum is 288MiB, assuming an average serialized block size of 2MB.**
```
```vim
:%s/288/275 /g
```

Per
```
$ src/qt/bitcoin-qt -maxuploadtarget=12 -regtest -printtoconsole|grep Recomm
2016-08-26 19:22:09 Max outbound target is very small (12 MiB) and will be overshot. Recommended minimum is 274.658 MiB.
```

But consider it a nit. Can be merged now.

luke-jr commented:
There are constantly 144 blocks per day. Segwit blocks *can* be up to 4 MB. If users set a value lower than this, what happens when that assumption is violated?

luke-jr commented on src/net.cpp:
```diff
@@ -2193,11 +2193,12 @@ void CNode::RecordBytesSent(uint64_t bytes)
     LOCK(cs_totalBytesSent);
-    uint64_t recommendedMinimum = (nMaxOutboundTimeframe / 600) * MAX_BLOCK_SERIALIZED_SIZE;
+    // assume the average size of serialized blocks is half the maximum
+    uint64_t recommendedMinimum = (nMaxOutboundTimeframe / 600) * MAX_BLOCK_SERIALIZED_SIZE / 2;
```
Suggest defining EXPECTED_AVERAGE_BLOCK_SIZE somewhere.

MarcoFalke commented:
@luke-jr IIRC, nothing will happen. You just stop to serve historical blocks earlier. I think you can even set `-maxuploadtarget=1` to effectively never send historical blocks.

gmaxwell commented:
uh. so the reason that it is a full day is because that is the reset interval.

Since the limiting will not prevent it from sending blocks, the purpose of the minimum was that it was the true minimum, and setting lower would pretty much guarantee overshoot.

MarcoFalke commented:
Though, `-maxuploadtarget` only applies to "historic blocks", so implying that it will not overshoot when you set it to *maximum block chain growth in one day* is misleading.
Regardless of the value that is set (may it be `0` or `100000`), there is no guarantee that it will not overshoot.

MarcoFalke commented:
So we might as well just remove the "recommended" minimum and call it done. Effectively you are still not helping IBD if you set the target to `144` or `288` instead of `0`.

jonasschnelli commented:
Closing in favor of #8712
